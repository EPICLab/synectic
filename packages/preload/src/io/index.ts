export * from './io-decompress';
export * from './io-exec';
export * from './io-extension';
export * from './io-extractName';
export * from './io-extractStats';
export * from './io-filterReadArray';
export * from './io-getDescendants';
export * from './io-isDescendant';
export * from './io-isDirectory';
export * from './io-isEqualPaths';
export * from './io-readDir';
export * from './io-readFile';
export * from './io-validateFileName';
export * from './io-writeFile';

/**
 * **WARNING** If a method is not included here, please see
 * [node-fs-extra](https://github.com/jprichardson/node-fs-extra) for possible utility functions
 * that improve upon the default filesystem functions in [Node.js FS
 * module](https://nodejs.org/api/fs.html). If there is no functions in `node-fs-extra` or `fs`,
 * then a new method can be added here.
 */
