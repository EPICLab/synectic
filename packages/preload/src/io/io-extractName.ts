import type {PathLike} from 'fs-extra';

/**
 * Extract the file basename from the path. Returns the filename, including extension, after the
 * last platform-agnostic file separator (e.g. '/src/api/router.asp' is translated to 'router.asp').
 * Exhibits different behavior from `path.basename` since that API uses `path.delimiter` in order to
 * set the delimiter for path processing, and this function is instead able to process paths across
 * platforms (e.g. processing Windows paths on a MacOS platform).
 *
 * @param filepath The relative or absolute path to evaluate.
 * @returns {string} A string containing the file basename.
 */
export const extractFilename = (filepath: PathLike): string => {
  const filename = filepath.toString().split(/[\\/]/).pop() as string;
  // filename can safely be cast as string because although pop() has a return type of string | undefined, it
  // cannot actually return undefined because split() returns string[] that is at worst empty
  return filename;
};

/**
 * Extract the directory name from the path. Returns the nearest directory name, excluding file
 * separators (e.g. '/src/api/router.asp' is translated to 'api'). Exhibits different behavior from
 * `path.dirname` since that API returns the full path of directories and includes '.' as a result
 * for a single directory path, whereas this function always returns only one directory name (or ''
 * if only a path to a file is provided).
 *
 * @param filepath The relative or absolute path to evaluate.
 * @returns {string} A string containing the directory name.
 */
export const extractDirname = (filepath: PathLike): string => {
  const trailingSeparator = filepath.toString().slice(-1)[0]?.match(/[\\/]/) !== null;
  const expandedPath = filepath.toString().split(/[\\/]/);
  if (expandedPath.length > 1) return expandedPath[expandedPath.length - 2] as string;
  if (trailingSeparator) return expandedPath[expandedPath.length - 1] as string;
  else return '';
};
