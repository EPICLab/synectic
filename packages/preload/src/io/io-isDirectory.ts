import {stat, type PathLike} from 'fs-extra';

/**
 * Asynchronously checks for the existence of a directory at the path.
 *
 * @param filepath The relative or absolute path to evaluate.
 * @returns {Promise<boolean>} A Promise object containing a boolean indicating true if the filepath
 *   is a directory, or false otherwise. Throws ENOENT error on non-existent filepath.
 */
export const isDirectory = async (filepath: PathLike): Promise<boolean> =>
  (await stat(filepath.toString())).isDirectory();
