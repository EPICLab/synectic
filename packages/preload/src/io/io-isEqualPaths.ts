import type {PathLike} from 'fs-extra';
import {relative} from 'path';

/**
 * Evaluates whether two paths are equivalent.
 *
 * @param path1 The relative or absolute path to evaluate.
 * @param path2 The other relative or absolute path to evaluate.
 * @returns {boolean} A boolean indicating true if the paths are equivalent, or false otherwise.
 */
export const isEqualPaths = (path1: PathLike, path2: PathLike): boolean => {
  return relative(path1.toString(), path2.toString()).length === 0;
};
