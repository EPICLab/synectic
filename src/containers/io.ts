import * as fs from 'fs-extra';
import * as path from 'path';

/**
 * Converts a JavaScript Object Notation (JSON) string into a typed object.
 * @param json A valid JSON string.
 * @return A typed object (or nested array of objects).
 */
export const deserialize = <T>(json: string) => {
  return JSON.parse(json) as T;
}

/**
 * Exract the file basename from the path. Returns the filename, including
 * extension, after the last platform-specific file separator.
 * @param filepath The relative or absolute path to evaluate.
 * @return A string containing the file basename.
 */
export const extractFilename = (filepath: fs.PathLike) => {
  return path.parse(filepath.toString()).base;
}

/**
 * Extract the file extension from the path. Returns the extension after 
 * the last period character in the path, otherwise returns full path if 
 * first character is a period or no period exists.
 * @param filepath The relative or absolute path to evaluate.
 * @return A string containing the file extension.
 */
export const extractExtension = (filepath: fs.PathLike) => {
  const ext = filepath.toString().split('.').pop();
  if (ext === undefined) return filepath.toString();
  else return ext;
}

/**
 * Asynchronously read file contents into a string.
 * @param filepath A valid filename or path to read from.
 * @return A Promise object for a string containing the file contents.
 */
export const readFileAsync = (filepath: fs.PathLike): Promise<string> => {
  return new Promise((resolve, reject) => {
    fs.readFile(path.resolve(filepath.toString()), (error, result) => {
      if (error) reject(error);
      else resolve(result.toString());
    });
  });
}

/**
 * Asynchronously read filenames contained within a target directory.
 * @param filepath A valid directory path to read from.
 * @return A Promise object for an array of strings containing filenames.
 */
export const readDirAsync = (filepath: fs.PathLike): Promise<string[]> => {
  return new Promise((resolve, reject) => {
    fs.readdir(path.resolve(filepath.toString()), (error, files) => {
      if (error) reject(error);
      else resolve(files);
    })
  });
};

/**
 * Asynchronously write data to a file. Creates a new file if none exists; will 
 * destructively rewrite existing files.
 * @param filepath A valid filename or path to write data to.
 * @param data A string containing content.
 */
export const writeFileAsync = (filepath: fs.PathLike, data: string): Promise<void> => {
  return new Promise((resolve, reject) => {
    fs.writeFile(path.resolve(filepath.toString()), data, (error) => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}