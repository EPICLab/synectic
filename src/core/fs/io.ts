import * as fs from 'fs-extra';
import * as path from 'path';

/**
 * Return the extension of the path, from the last '.' to end of string in the last portion of the path.
 * If there is no '.' in the last portion of the path or the first character of it is '.', then it returns the entire string.
 * @param p The path to evaluate.
 */
export function extname(p: string): string {
  let ext: string | undefined = p.split('.').pop();
  if (ext !== undefined) return ext;
  else return p;
}

/**
 * Converts a JSON string into an object.
 * @param json A valid JSON string.
 * @return An object (can be an array of objects).
 */
export function deserialize<T>(json: string): T {
  return <T> JSON.parse(json);
}

/**
 * Asynchronously reads file content into a string.
 * @param filename A valid filename or path to read from.
 * @return A string containing the file content.
 */
export function readFileAsync(filename: string): Promise<string> {
  return new Promise((resolve, reject) => {
    fs.readFile(path.resolve(filename), (error, result) => {
      if (error) reject(error);
      else resolve(result.toString());
    });
  });
}

/**
 * Asynchronously writes to a file; creates a new file if none exists.
 * @param filename A valid filename or path to write the data to.
 * @param data A string containing content.
 */
export function writeFileAsync(filename: string, data: string): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.writeFile(path.resolve(filename), data, path.extname(filename), (error) => {
      if (error) reject(error);
      else {
        console.info('File `' + path.resolve(filename) + '` created.');
        resolve();
      };
    });
  });
}
