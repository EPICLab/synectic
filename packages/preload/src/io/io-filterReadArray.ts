import type {PathLike} from 'fs-extra';
import {isDirectory} from './io-isDirectory';

/**
 * Asynchronously filter for either directories or files from an array of both. Returns filepaths of
 * all child directories (default), or all child files if `fileOnly` option is enabled.
 *
 * @param filepaths Array containing filepaths for directories and files.
 * @param fileOnly Flag for returning only filepaths for files; defaults to false.
 * @returns {Promise<PathLike[]>} A Promise object for an array containing filepaths for either all
 *   child directories or all child files.
 */

export const filterReadArray = async (
  filepaths: PathLike[],
  fileOnly = false,
): Promise<PathLike[]> => {
  return await filepaths.reduce(
    async (previousPromise: Promise<PathLike[]>, filepath: PathLike) => {
      const collection = await previousPromise;
      const directory = await isDirectory(filepath);
      if (fileOnly && !directory) collection.push(filepath);
      if (!fileOnly && directory) collection.push(filepath);
      return collection;
    },
    Promise.resolve([]),
  );
};
