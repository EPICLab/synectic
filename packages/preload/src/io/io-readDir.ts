import {type PathLike, readdir} from 'fs-extra';
import {join, resolve as resolvePath} from 'path';
import {isDirectory} from './io-isDirectory';
import {flattenArray} from '../flatten';

/**
 * Asynchronously read filenames contained within a target directory.
 *
 * @param filepath A valid directory path to read from.
 * @returns {Promise<string[]>} A Promise object for an array of filenames. Throws ENOENT error on
 *   non-existent filepath, and ENOTDIR error on a filepath pointing to a file.
 */
export const readDirAsync = (filepath: PathLike): Promise<string[]> => {
  return new Promise<string[]>((resolve, reject) => {
    readdir(resolvePath(filepath.toString()), (error, files) => {
      if (error) reject(error);
      else resolve(files);
    });
  });
};

/**
 * Asynchronously and recursively descends from a root directory to read filenames and child
 * directories. Descends to `depth` level, if specified, otherwise defaults to recursively visiting
 * all sub-directories.
 *
 * @param filepath A valid directory path to read from.
 * @param depth Number of sub-directories to descend; defaults to infinity.
 * @returns {Promise<string[]>} A Promise object for an array of filenames. Throws ENOENT error on
 *   non-existent filepath, and ENOTDIR error on a filepath pointing to a file.
 */
export const readDirAsyncDepth = async (
  filepath: PathLike,
  depth = Infinity,
): Promise<string[]> => {
  const files = await Promise.all(
    (
      await readdir(filepath.toString())
    ).map(async f => {
      const fullPath = join(filepath.toString(), f);
      return depth > 1 && (await isDirectory(fullPath))
        ? await readDirAsyncDepth(fullPath, depth - 1)
        : fullPath;
    }),
  );
  return [...flattenArray(files), filepath.toString()];
};
