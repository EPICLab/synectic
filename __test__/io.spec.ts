import mock from 'mock-fs';
import * as io from '../src/containers/io';
import * as fs from 'fs-extra';

describe('io.deserialize', () => {
  it('deserialize to parse a JSON string into a TypeScript object', () => {
    const json = '{"result":true, "count":42}';
    type typedJson = { result: boolean; count: number };
    const deserializedJson = io.deserialize<typedJson>(json);
    expect(typeof deserializedJson.result).toBe('boolean');
    expect(deserializedJson.result).toBe(true);
    expect(deserializedJson).toMatchSnapshot();
  });

  it('deserialize fails with an error on malformed JSON', () => {
    // eslint-disable-next-line no-useless-escape
    const malformedJson = '{ "key": "Something \\\\"Name\\\\" something\", "anotherkey": "value" }';
    expect(() => io.deserialize(malformedJson)).toThrow(SyntaxError);
  });
});

describe('io.extractStats', () => {

  beforeAll(() => {
    mock({
      foo: {
        bar: mock.file({
          content: 'file contents',
          ctime: new Date(1),
          mtime: new Date(1)
        })
      }
    });
  });

  afterAll(mock.restore);

  it('extractStats to extract relevant file information from valid path', async () => {
    return expect(io.extractStats('foo/bar')).resolves.toHaveProperty('ctime', new Date(1));
  });

  it('extractStats to return undefined from nonexistent path', async () => {
    return expect(io.extractStats('foo/baz')).resolves.toBeUndefined();
  });
});

describe('io.extractFilename', () => {
  it('extractFilename to extract filename from Linux/MacOS paths', () => {
    expect(io.extractFilename('/Users/foo/bar/module.d.ts')).toBe('module.d.ts');
    expect(io.extractFilename('./baz/webpack.config.js')).toBe('webpack.config.js');
    expect(io.extractFilename('../../baz/sample.c9search_results')).toBe('sample.c9search_results');
  });

  it('extractFilename to extract filename from Windows paths', () => {
    expect(io.extractFilename('C:\\Foo\\Bar\\Baz\\file.js')).toBe('file.js');
    expect(io.extractFilename('2018\\January.xlsx')).toBe('January.xlsx');
  });
});

describe('io.extractDirname', () => {
  it('extractDirname to extract dirname from Linux/MacOS paths', () => {
    expect(io.extractDirname('/Users/foo/bar/module.d.ts')).toBe('bar');
    expect(io.extractDirname('./baz/webpack.config.js')).toBe('baz');
    expect(io.extractDirname('../../baz/sample.c9search_results')).toBe('baz');
    expect(io.extractDirname('/Users/foo/bar/')).toBe('bar');
    expect(io.extractDirname('bar/')).toBe('bar');
    expect(io.extractDirname('/setup.cfg')).toBe('');
    expect(io.extractDirname('module.d.ts')).toBe('');
  });

  it('extractDirname to extract dirname from Windows paths', () => {
    expect(io.extractDirname('C:\\Foo\\Bar\\Baz\\file.js')).toBe('Baz');
    expect(io.extractDirname('2018\\January.xlsx')).toBe('2018');
    expect(io.extractDirname('C:\\Foo\\Bar\\Baz\\')).toBe('Baz');
  });

  it('extractDirname resolves malformed paths', () => {
    expect(io.extractDirname('')).toBe('');
    expect(io.extractDirname('/')).toBe('');
    expect(io.extractDirname(' /')).toBe(' ');
    expect(io.extractDirname('/ ')).toBe('');
  });
});

describe('io.extractExtension', () => {
  it('extractExtension to extract extension from filename', () => {
    expect(io.extractExtension('foo.js')).toBe('js');
    expect(io.extractExtension('bar.d.ts')).toBe('ts');
    expect(io.extractExtension('.htaccess')).toBe('htaccess');
  });

  it('extractExtension to extract extension from Linux/MacOS paths', () => {
    expect(io.extractExtension('/Users/foo/bar/module.d.ts')).toBe('ts');
    expect(io.extractExtension('./baz/webpack.config.js')).toBe('js');
    expect(io.extractExtension('../../baz/sample.c9search_results')).toBe('c9search_results');
  });

  it('extractExtension to extract extension from Windows paths', () => {
    expect(io.extractExtension('C:\\Foo\\Bar\\Baz\\file.js')).toBe('js');
    expect(io.extractExtension('2018\\January.xlsx')).toBe('xlsx');
  });
});

describe('io.readFileAsync', () => {
  beforeAll(() => {
    mock({
      'foo/bar': {
        'some-file.txt': 'file contents',
        'empty-dir': {/** empty directory */ }
      },
      'baz/qux': {
        'some.png': Buffer.from([8, 6, 7, 5, 3, 0, 9]),
        'vex/bol/wiz': {/** another empty directory */ }
      }
    });
  });

  afterAll(mock.restore);

  it('readFileAsync resolves to string for text file content', async () => {
    await expect(io.readFileAsync('foo/bar/some-file.txt', { encoding: 'utf8' })).resolves.toBe('file contents');
  });

  it('readFileAsync resolves to Buffer for bytestring file content', async () => {
    await expect(io.readFileAsync('baz/qux/some.png')).resolves.toEqual(Buffer.from([8, 6, 7, 5, 3, 0, 9]));
  });

  it('readFileAsync fails with error on non-existing file', async () => {
    await expect(io.readFileAsync('foo/bar/empty-dir/nonexist.js')).rejects.toThrow(/ENOENT/);
  });
});

describe('io.readDirAsync', () => {
  beforeAll(() => {
    mock({
      'foo/bar': {
        'some-file.txt': 'file contents',
        'empty-dir': {/** empty directory */ }
      },
      'baz/qux': {
        'nup/tul/some.png': Buffer.from([8, 6, 7, 5, 3, 0, 9]),
        'vex/bol/wiz': {/** another empty directory */ }
      }
    });
  });

  afterAll(mock.restore);

  it('readDirAsync to resolve to array of child filepaths', async () => {
    await expect(io.readDirAsync('foo/bar')).resolves.toHaveLength(2);
  });

  it('readDirAsync fails with an error', async () => {
    await expect(io.readDirAsync('foo/dep/')).rejects.toThrow(/ENOENT/);
  });
});

describe('io.isDirectory', () => {
  beforeAll(() => {
    mock({
      foo: {
        bar: mock.file({ content: 'file contents', ctime: new Date(1) }),
        zap: {
          zip: mock.file({ content: 'file contents', ctime: new Date(1) }),
        }
      },
      empty: {},
    });
  });

  afterAll(mock.restore);

  it('isDirectory resolves to true for empty directories', () => {
    return expect(io.isDirectory('empty')).resolves.toBe(true);
  });

  it('isDirectory resolves to true for non-empty directories', () => {
    return expect(io.isDirectory('foo/zap')).resolves.toBe(true);
  });

  it('isDirectory resolves to false for file', () => {
    return expect(io.isDirectory('foo/bar')).resolves.toBe(false);
  });
});

describe('io.readDirAsyncDeep', () => {
  beforeAll(() => {
    mock({
      foo: {
        bar: mock.file({ content: 'file contents', ctime: new Date(1) }),
        baz: mock.file({ content: 'file contents', ctime: new Date(1) }),
        zap: {
          zed: {
            beq: mock.file({ content: 'file contents', ctime: new Date(1) }),
            bup: mock.file({ content: 'file contents', ctime: new Date(1) })
          },
          zip: mock.file({ content: 'file contents', ctime: new Date(1) }),
        }
      },
      zonk: {
        zork: mock.file({ content: 'file contents', ctime: new Date(1) }),
      },
      imp: {
        bamp: {},
      },
      empty: {},
    });
  });

  afterAll(mock.restore);

  it('readDirAsyncDeep resolves an empty directory', () => {
    return expect(io.readDirAsyncDeep('empty')).resolves.toHaveLength(1);
  });

  it('readDirAsyncDeep resolves a directory with sub-files', () => {
    return expect(io.readDirAsyncDeep('zonk')).resolves.toHaveLength(2);
  });

  it('readDirAsyncDeep resolves a directory with sub-directories', () => {
    return expect(io.readDirAsyncDeep('imp')).resolves.toHaveLength(2);
  });

  it('readDirAsyncDeep inclusively resolves a directory with multiple layers of directories and files', () => {
    return expect(io.readDirAsyncDeep('foo', true)).resolves.toHaveLength(8);
  });

  it('readDirAsyncDeep exclusively resolves a directory with multiple layers of directories and files', () => {
    return expect(io.readDirAsyncDeep('foo', false)).resolves.toHaveLength(7);
  });

  it('readDirAsyncDeep fails with an error on non-existent paths', () => {
    return expect(io.readDirAsyncDeep('foo/dep/')).rejects.toThrow(/ENOENT/);
  });
});

describe('io.writeFileAsync', () => {
  beforeAll(() => {
    mock({
      'foo/bar': {
        'fileB.txt': 'version 1'
      },
      'baz': {/** empty directory */ }
    });
  });

  afterAll(mock.restore);

  it('writeFileAsync to resolve and write a new file with content', async () => {
    const testPath = 'foo/bar/fileA.txt';
    await io.writeFileAsync(testPath, 'sample data');
    await expect(fs.ensureFile(testPath)).resolves.not.toThrow();
    await expect(io.readFileAsync(testPath, { encoding: 'utf8' })).resolves.toBe('sample data');
  });

  it('writeFileAsync to resolve and overwrite an existing file with content', async () => {
    const testPath = 'foo/bar/fileB.txt';
    await expect(io.readFileAsync(testPath, { encoding: 'utf8' })).resolves.toBe('version 1');
    await io.writeFileAsync(testPath, 'version 2');
    await expect(io.readFileAsync(testPath, { encoding: 'utf8' })).resolves.toBe('version 2');
  });
});