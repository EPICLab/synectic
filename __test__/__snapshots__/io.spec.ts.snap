// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`io.deserialize deserialize to parse a JSON string into a TypeScript object 1`] = `
Object {
  "count": 42,
  "result": true,
}
`;
